task_processing/plugins/mesos/execution_framework.py

```diff
@@ -481,110 +282,109 @@ def create_new_docker_task(self, offer, task_config):
+                driver.declineOffer(offer.id, self.offer_decline_filter)
+                continue
+
+            log.info("Launching {number} new docker task(s) using offer {id} on slave {slave}".format(
+                number=len(tasks_to_launch),
+                id=offer.id.value,
+                slave=offer.slave_id.value
             )
-            for offer in offers:
-                driver.declineOffer(offer.id)
-        else:
-            for idx, offer in enumerate(offers):
-                if offer.slave_id.value in self.blacklisted_slaves:
-                    log.critical(
-                        'Ignoring offer {0} from blacklisted slave {1}'.format(
-                            offer.id.value, offer.slave_id.value
-                        )
-                    )
-                    driver.declineOffer(offer.id)
-                    del offers[idx]
-
-            if len(offers) == 0:
-                return
-
-            self.launch_tasks_for_offers(driver, offers)
+            )
+            driver.launchTasks(offer.id, tasks_to_launch)
 
-    def statusUpdate(self, driver, update):
-        if self.queue:
-            self.queue.put(self.translator(update))
-        if self.frozen:
-            return
+            for task in tasks_to_launch:
+                self.task_metadata[task.task_id.value]['retries'] += 1
+                self.task_metadata[task.task_id.value]['time_launched'] = time.time(
+                )
+                self.task_metadata[task.task_id.value]['task_state'] = 'launched'
 
-        # update tasks
+    def statusUpdate(self, driver, update):
         task_id = update.task_id.value
-        state = update.state
-        task_params = self.tasks_with_flags.setdefault(
-            task_id, MesosTaskParameters(health=None))
-        task_params.mesos_task_state = state
-
-        for task, params in list(self.tasks_with_flags.items()):
-            if params.marked_for_gc:
-                self.tasks_with_flags.pop(task)
-
-        if task_params.mesos_task_state is not TASK_STAGING:
-            if self.tasks_with_flags[task_id].staging_timer:
-                self.tasks_with_flags[task_id].staging_timer.cancel()
-                self.tasks_with_flags[task_id].staging_timer = None
-
-        # if task_params.mesos_task_state not in LIVE_TASK_STATES:
-            # task_params.marked_for_gc = True
-            # with self.constraint_state_lock:
-                # update_constraint_state(task_params.offer, self.constraints,
-                # self.constraint_state, step=-1)
+        log.info("Task update {update} received for task {task}".format(
+            update=mesos_pb2.TaskState.Name(update.state),
+            task=task_id
+        )
+        )
 
+        if update.state == mesos_pb2.TASK_RUNNING:
+            self.task_metadata[task_id]['task_state'] = 'in-flight'
+
+        if update.state == mesos_pb2.TASK_FINISHED:
+            self.task_update_queue.put(self.translator(update))
```

Comment:
something like retry_policy would be a task_executor-level option, not execution framework imo. this way if somebody doesn't like our executor they can implement another one and be able to deal with all events without having to touch mesos-related code. lets just publish all events into that queue and have executor/runner deal with them (and deal with retries).

execution framework would only be a place for dealing with mesos offers, reconciliations, failover, etc.